src/fu-engine.c

```diff
@@ -2999,6 +2999,47 @@ fu_engine_schedule_update(FuEngine *self,
 	return fu_engine_offline_setup(error);
 }
 
+static gboolean
+fu_engine_retain_firmware_archive(FuEngine *self, GBytes *fw, GError **error)
+{
+	FwupdRemote *remote_tmp = fu_remote_list_get_by_id(self->remote_list, "backup");
+	g_autofree gchar *localstatepkg = fu_path_from_kind(FU_PATH_KIND_LOCALSTATEDIR_PKG);
+	g_autofree gchar *backupdir = g_build_filename(localstatepkg, "backup", NULL);
+	g_autofree gchar *backupdir_uri = g_strdup_printf("file://%s", backupdir);
+	g_autofree gchar *remotes_path = fu_path_from_kind(FU_PATH_KIND_LOCALSTATEDIR_REMOTES);
+	g_autofree gchar *remotes_fn = g_build_filename(remotes_path, "backup.conf", NULL);
+	g_autofree gchar *archive_checksum = g_compute_checksum_for_bytes(G_CHECKSUM_SHA256, fw);
+	g_autofree gchar *archive_basename = g_strdup_printf("%s.cab", archive_checksum);
+	g_autofree gchar *archive_fn = g_build_filename(backupdir, archive_basename, NULL);
+	g_autoptr(FwupdRemote) remote = fwupd_remote_new();
+
+	/* save archive if required */
+	if (!g_file_test(archive_fn, G_FILE_TEST_EXISTS)) {
+		g_info("saving archive to %s", archive_fn);
+		if (!fu_bytes_set_contents(archive_fn, fw, error))
+			return FALSE;
+	}
+
+	/* already exists as an enabled remote */
+	if (remote_tmp != NULL && fwupd_remote_get_enabled(remote_tmp))
+		return TRUE;
+
+	/* just enable */
+	if (remote_tmp != NULL) {
+		g_debug("enabling remote %s", fwupd_remote_get_id(remote_tmp));
```

Comment:
with the new debugging stuff shouldn't this be `g_info`?

Reply:
It should indeed. Hoisted by my own petard. Fixed.